File: src/cg.cc
#include "cg.h"
#include "utils.h"

using namespace arma;

void CG::solve(const mat &A, const vec &b, vec &x,
    const bool restart, const size_t iterations) {

    if (restart) {
        r = A * x - b;
        p = -r;
        prev_r_sq_sum = dot(r, r);
    }
    
    for (size_t i = 0; i < iterations; i++){
        const vec Ap = A * p;
        const double alpha = prev_r_sq_sum / dot(p, Ap);

        x += alpha * p;
        r += alpha * Ap;
        const double r_sq_sum = dot(r, r);

        const double beta = r_sq_sum / prev_r_sq_sum;
        p *= beta;
        p -= r;
        prev_r_sq_sum = r_sq_sum;
    }

}

File: src/cg.m
function [x, p, r] = cg(A, b, x, restart, iters, p, r)
    
    if restart == 1
        r = A * x - b;
        p = -r;
    end
    
    for i=1:iters
        Ap = A * p;
        old_r_sum = dot(r, r);
        if old_r_sum <= 1e-12
            break;
        end
        alpha = old_r_sum / (p' * Ap);

        x = x + alpha * p;
        r = r + alpha * Ap;

        be = (r' * r) / old_r_sum;
        p = -r + be * p;

    end
end